share/sysbuild/cmake/modules/sysbuild_extensions.cmake

```diff
@@ -2,6 +2,83 @@
+#
+# This function will load the CMakeCache.txt file from the binary directory
+# given by the BINARY_DIR argument.
+#
+# All CMake cache variables are stored in a custom target which is identified by
+# the name given as value to the IMAGE argument.
+#
+# IMAGE:      image name identifying the cache for later sysbuild_get() lookup calls.
+# BINARY_DIR: binary directory (build dir) containing the CMakeCache.txt file to load.
+function(load_cache)
+  set(single_args APPLICATION BINARY_DIR)
+  cmake_parse_arguments(LOAD_CACHE "" "${single_args}" "" ${ARGN})
+
+  if(NOT TARGET ${LOAD_CACHE_APPLICATION}_cache)
+    add_custom_target(${LOAD_CACHE_APPLICATION}_cache)
+  endif()
+  file(STRINGS "${LOAD_CACHE_BINARY_DIR}/CMakeCache.txt" cache_strings)
+  foreach(str ${cache_strings})
+    # Using a regex for matching whole 'VAR_NAME:TYPE=VALUE' will strip semi-colons
+    # thus resulting in lists to become strings.
+    # Therefore we first fetch VAR_NAME and TYPE, and afterwards extract
+    # remaining of string into a value that populates the property.
+    # This method ensures that both quoted values and ;-separated list stays intact.
+    string(REGEX MATCH "([^:]*):([^=]*)=" variable_identifier ${str})
+    if(NOT "${variable_identifier}" STREQUAL "")
+      string(LENGTH ${variable_identifier} variable_identifier_length)
+      string(SUBSTRING "${str}" ${variable_identifier_length} -1 variable_value)
+      set_property(TARGET ${LOAD_CACHE_APPLICATION}_cache APPEND PROPERTY "CACHE:VARIABLES" "${CMAKE_MATCH_1}")
+      set_property(TARGET ${LOAD_CACHE_APPLICATION}_cache PROPERTY "${CMAKE_MATCH_1}:TYPE" "${CMAKE_MATCH_2}")
+      set_property(TARGET ${LOAD_CACHE_APPLICATION}_cache PROPERTY "${CMAKE_MATCH_1}" "${variable_value}")
+    endif()
+  endforeach()
+endfunction()
+
+# Usage:
+#   sysbuild_get(<variable> IMAGE <image> [VAR <image-variable>])
+#
+# This function will return the variable found in the CMakeCache.txt file
+# identified by image.
+# If `VAR` is provided, the name given as parameter will be looked up, but if
+# `VAR` is not given, the `<variable>` name provided will be used both for
+# lookup and value return.
+#
+# The result will be returned in `<variable>`.
+#
+# Example use:
+#   sysbuild_get(PROJECT_NAME IMAGE my_sample)
+#     will lookup PROJECT_NAME from the CMakeCache identified by `my_sample` and
+#     and return the value in the local variable `PROJECT_NAME`.
+#
+#   sysbuild_get(my_sample_PROJECT_NAME IMAGE my_sample VAR PROJECT_NAME)
+#     will lookup PROJECT_NAME from the CMakeCache identified by `my_sample` and
+#     and return the value in the local variable `my_sample_PROJECT_NAME`.
+#
+# <variable>: variable used for returning CMake cache value. Also used as lookup
+#             variable if `VAR` is not provided.
+# IMAGE:      image name identifying the cache to use for variable lookup.
+# VAR:        name of the CMake cache variable name to lookup.
+function(sysbuild_get variable)
```

Comment:
This does not empty the input variable if the VAR variable does not exist in the target. E.g. if the input variable holds "test" and VAR is set to "BOB" which doesn't exist, the input variable will still contain "test"

Reply:
> This does not empty the input variable if the VAR variable does not exist in the target. E.g. if the input variable holds "test" and VAR is set to "BOB" which doesn't exist, the input variable will still contain "test"

That is correct.
In this case it's actually the callers responsibility to ensure the var is undefined.
A couple of reasons:
- We cannot use the general CMake approach of `<var>-NOTFOUND` because a cache var could contain the value `var-NOTFOUND` making it impossible to differentiate the two case / as to why `-NOTFOUND` was returned.
- Clearing the var using `unset(<var> PARENT_SCOPE)` would only clear the normal var, but the var could actually be a cache var, and thus still be present. And we should not clear sysbuild cache vars in such cases.
- We cannot set the var to empty string, cause the real cache var could actually be an empty string.

so basically it's the callers responsibility to ensure the var is undefined, if that is important to know.

What we can do is to print a warning if the return var is already defined with a value, and maybe even hint the caller that `sysbuild_get(<var> ... VAR <imager-variable>)` should be used instead.

Reply:
Warning printed if caller is using an already defined variable.